sycl/doc/extensions/proposed/sycl_ext_oneapi_kernel_compiler.asciidoc

```diff
@@ -0,0 +1,501 @@
+----
+!====
+
+Overload (1):
+
+_Constraints:_ Available only when `PropertyListT` is an instance of
+`sycl::ext::oneapi::experimental::properties` which contains no properties
+other than those listed below in the section "New properties".
+
+_Effects:_ The source code from `sourceBundle` is translated into one or more
+device images of state `bundle_state::executable`, and a new kernel bundle is
+created to contain these device images.
+The new bundle represents all of the kernels in `sourceBundle` that are
+compatible with at least one of the devices in `devs`.
+Any remaining kernels (those that are not compatible with any of the devices in
+`devs`) are not represented in the new kernel bundle.
+
+The new bundle has the same associated context as `sourceBundle`, and the new
+bundle's set of associated devices is `devs` (with duplicate devices removed).
+
+_Returns:_ The newly created kernel bundle, which has `executable` state.
+
+_Throws:_
+
+* An `exception` with the `errc::invalid` error code if any of the devices in
+  `devs` is not contained by the context associated with `sourceBundle`.
+
+* An `exception` with the `errc::invalid` error code if `props` contains an
+  `options` property that specifies an invalid option.
+
+* An `exception` with the `errc::build` error code if the compilation or
+  linking operations fail.
+  In this case, the exception `what` string contains a description of the
+  error.
+  This string is intended for human consumption, and the format may not be
+  stable across implementations of this extension.
+
+Overload (2):
+
+_Constraints:_ Same as overload (1).
+
+_Effects:_ Equivalent to `build(sourceBundle, ctxt.get_devices(), props)`.
+|====
+
+=== New properties
+
+This extension adds the following properties, which can be used in conjunction
+with the `build` function that is defined above:
+
+|====
+a|
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+namespace sycl::ext::oneapi::experimental {
+
+struct build_options {
+  std::vector<std::string> opts;
+  build_options(const std::string &name);               (1)
+  build_options(const std::vector<std::string> &opts);  (2)
+};
+using build_options_key = build_options;
+
+template<>
+struct is_property_key<build_options_key> : std::true_type {};
+
+} // namespace sycl::ext::oneapi::experimental {
+----
+!====
+
+This property provides build options that may affect the compilation or linking
+of the kernel, where each build option is a string.
+There are no standard build options that are common across all source
+languages.
+Instead, each source language specification defines its own set of build
+options.
+
+_Effects (1):_ Constructs a `build_options` property with a single build
+option.
+
+_Effects (2):_ Constructs a `build_options` property from a vector of build
+options.
+
+a|
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+namespace sycl::ext::oneapi::experimental {
+
+struct build_log {
+  std::string *log;
+  build_log(std::string *log);  (1)
```

Comment:
How about `build_log(result)`?

I think the names for most (all?) of our properties are nouns (not verbs).  Is that a good reason to stick with that convention?

Reply:
`build_log(result)` is an improvement, at least from a specification readability perspective.

I agree we mostly use nouns, but SYCL 2020 does have some verbs (e.g., `use_mutex`).

Could we sidestep this issue by just putting all of the information in the exception, instead?  If we're just trying to avoid the log showing up in `what()`, would it make sense to have a special `sycl::build_exception` with a `get_build_log()` function?

Reply:
The `build_log` property is only useful for a `build` that succeeds.  A failing build does throw an exception and the log message is in the "what" string.  The `build_log` is only useful for getting compiler information (e.g. warning or optimization report) from a build that does not fail.

Reply:
Oh, I'm sorry I missed that.  I somehow focused in on this part:

> When a build fails, an `exception` is thrown and the exception's `what` string provides a description of the error.

...and interpreted "a description" as being different to the complete build log.  I imagined high-level exception messages like "Error in source code." or "Unrecognized build options."  I think it would be slightly clearer if this said something like:

> When a build fails, an `exception` is thrown and the exception's `what` string provides a full build log, including descriptions of any errors, warning messages and other diagnostics.

I think we should also add a non-normative note that says something like:

> Note: An uncaught exception may result in some or all of the source code used to create the `kernel_bundle` being printed to the terminal. In situations where this is undesirable, developers must ensure that the exception is caught and handled appropriately.

Reply:
Yes, I can clarify the wording around the case when `build` fails due to a compilation error, and I can add the non-normative note.

Thinking about the name of the `build_log` property ... the names `build_log_location` and `save_build_log` both seem pretty long.  How about `save_log(location)`?  I think it's clear that it's a "build" log because the property can only be passed to the `build` function.

Reply:
I'm okay with `save_log`.  It might even be better that it doesn't have "build" in the name, because we could choose to re-use it if we introduce other functionality later which also writes a log.

Reply:
Done in a4d29bc25be3592c27f70b3bcb802c6558c46016